### ISSUE TITLE
Xpath matcher poor performance

### ISSUE THREAD

**Issue by philippe-granet** [tomakehurst/wiremock#1048]
Wiremock version: 2.19.0

First of all, thank you for your work on wiremock, a very useful tool!

We are using wiremock for stubbing SOAP services in development environment.
We also tried to use Wiremock for performance tests, but there is some bottleneck:
* **javax.xml.transform.TransformerFactory.newInstance()**
This factory is instanciate everywhere in XMLUnit library and WireMock, but algorithm  to determine the implementation class to load is heavy:
https://docs.oracle.com/javase/7/docs/api/javax/xml/transform/TransformerFactory.html#newInstance()

Simply set system property `-Djavax.xml.transform.TransformerFactory=<implementation to load>` resolve this bottleneck

* **javax.xml.xpath.XPathFactory.newInstance()**
This factory is instanciate everywhere in XMLUnit library and WireMock, but algorithm  to determine the implementation class to load is heavy:
https://docs.oracle.com/javase/7/docs/api/javax/xml/xpath/XPathFactory.html#newInstance()
https://github.com/openjdk-mirror/jdk7u-jaxp/blob/master/src/javax/xml/xpath/XPathFactoryFinder.java#L156

Simply set system property `-Djavax.xml.xpath.XPathFactory:http://java.sun.com/jaxp/xpath/dom=<implementation to load>` resolve this bottleneck

* **XMLUnit.setXPathFactory** and **XMLUnit.setTransformerFactory**
Predefine theses factories in XMLUnit also improve performance
 
**Simple test case:**
```java
@Test
public void perfXmlTest() {
	perfRunDatas("without optimisations");

	// Optimisation 1
	String xPathFactoryImpl = XPathFactory.newInstance().getClass().getName();
	System.setProperty(XPathFactory.DEFAULT_PROPERTY_NAME + ":" 
		+ XPathFactory.DEFAULT_OBJECT_MODEL_URI, xPathFactoryImpl);
	perfRunDatas("with XPathFactory system property");

	// Optimisation 2
	String transformerFactoryImpl = TransformerFactory.newInstance().getClass().getName();
	System.setProperty(TransformerFactory.class.getName(), transformerFactoryImpl);
	perfRunDatas("with XPathFactory/TransformerFactory system properties");

	// Optimisation 3
	XMLUnit.setTransformerFactory(transformerFactoryImpl);
	perfRunDatas("with XPathFactory/TransformerFactory system properties and TransformerFactory predefined in XMLUnit");

	// Optimisation 4
	XMLUnit.setXPathFactory(xPathFactoryImpl);
	perfRunDatas("with XPathFactory/TransformerFactory system properties and TransformerFactory/XPathFactory predefined in XMLUnit");
}
private void perfRunDatas(String string) {
	long ts = System.currentTimeMillis();
	for (int i = 0; i < 10000; i++) {
		WireMock.matchingXPath("//perf-test").match("<perf-test>perf-data</perf-test>");
	}
	System.out.println(string + ":\n" + (System.currentTimeMillis() - ts) + "ms");
}
```
generate this:
**without optimisations:**
23745ms
**with XPathFactory system property:**
12690ms
**with XPathFactory/TransformerFactory system properties:**
2032ms
**with XPathFactory/TransformerFactory system properties and TransformerFactory predefined in XMLUnit:**
1981ms
**with XPathFactory/TransformerFactory system properties and TransformerFactory/XPathFactory predefined in XMLUnit:**
1767ms

So, simply adding to system properties:
`-Djavax.xml.transform.TransformerFactory=<implementation>`
`-Djavax.xml.xpath.XPathFactory:http://java.sun.com/jaxp/xpath/dom=<implementation>`
and predefining factories in XMLUnit library improved performance by 10 !

I think this should be described in documentation, or implemented in Wiremock.
An implementation could be calling this method `optimizeFactoriesLoading()` during WireMock initialisation :

```java
public void optimizeFactoriesLoading() {
	
	String transformerFactoryImpl = TransformerFactory.newInstance().getClass().getName();
	String xPathFactoryImpl = XPathFactory.newInstance().getClass().getName();
	
	setProperty(TransformerFactory.class.getName(), transformerFactoryImpl);
	setProperty(XPathFactory.DEFAULT_PROPERTY_NAME + ":" + XPathFactory.DEFAULT_OBJECT_MODEL_URI,
		xPathFactoryImpl);

	XMLUnit.setTransformerFactory(transformerFactoryImpl);
	XMLUnit.setXPathFactory(xPathFactoryImpl);
}
private String setProperty(final String name, final String value) {
	return AccessController.doPrivileged(new PrivilegedAction<String>() {
		@Override
		public String run() {
			return System.setProperty(name, value);
		}
	});
}

```

**Comment by tomakehurst** [comment]
Thanks for sharing this investigation, this is very useful to know.

I think the way forward is probably to try to detect the presence of the most common implementation(s) and set that as the system param on startup. Then only fall back to the default (dynamic detection) behaviour if none of the common impls are present (which I suspect might be the case in Android).

Can you see any obvious issues with this approach?

**Comment by tomakehurst** [comment]
Please ignore my last comment. Realised I hadn't read your code example properly!

**Comment by tomakehurst** [comment]
I've added your suggested code (please see the ref'd commit)

**Comment by philippe-granet** [comment]
Thanks! I close this issue (fixed in 50c07ef)
